Packages/ChatSecure/ChatSecure/Classes/WebRTCJanus/JanusRolePublish.swift

```diff
@@ -0,0 +1,395 @@
+			if videoCapturer != nil,
+			   let publishConstraints = mediaConstraints as? JanusPublishMediaConstraints {
+				startCaptureLocalVideo(cameraPositon: cameraDevicePosition,
+									   videoWidth: Int(publishConstraints.resolution.width),
+									   videoHeight: Int(publishConstraints.resolution.height),
+									   videoFps: Int(publishConstraints.fps))
+				//                startCaptureLocalVideo(cameraPositon: self.cameraDevicePosition,
+				//                                       videoWidth: 1280,
+				//                                       videoHeight: 720,
+				//                                       videoFps: 60)
+			}
+			localVideoTrack?.add(videoRenderView)
+		}
+	}
+	
+	func setupLocalViewFrame(frame: CGRect) {
+		videoRenderView?.frame = frame
+	}
+	
+	func captureCurrentFrame(sampleBuffer: CMSampleBuffer) {
+		if let capturer = videoCapturer as? RTCCustomFrameCapturer {
+			capturer.capture(sampleBuffer)
+		}
+	}
+	
+	func captureCurrentFrame(sampleBuffer: CVPixelBuffer) {
+		if let capturer = videoCapturer as? RTCCustomFrameCapturer {
+			capturer.capture(sampleBuffer)
+		}
+	}
+	
+	// MARK: Data Channels
+	private func createDataChannel() -> RTCDataChannel? {
+		let config = RTCDataChannelConfiguration()
+		guard let dataChannel = peerConnection.dataChannel(forLabel: "WebRTCData", configuration: config) else {
+			debugPrint("Warning: Couldn't create data channel.")
+			return nil
+		}
+		return dataChannel
+	}
+	
+	func sendData(_ data: Data) {
+		let buffer = RTCDataBuffer(data: data, isBinary: true)
+		self.remoteDataChannel?.sendData(buffer)
+	}
+	
+	func setupPeerStream() {
+		// Video
+		if channels.video, let videoTrack = localVideoTrack {
+			peerConnection.add(videoTrack, streamIds: ["stream0"])
+		}
+		// Audio
+		if channels.audio, let audioTrack = localAudioTrack {
+			peerConnection.add(audioTrack, streamIds: ["stream0"])
+		}
+		// Data
+		if channels.datachannel,
+		   let dataChannel = createDataChannel() {
+			dataChannel.delegate = self
+			localDataChannel = dataChannel
+		}
+	}
+	
+	// MARK: - Override function
+	override func joinRoom(withRoomId roomId: Int64, username: String?, callback: @escaping RoleJoinRoomCallback) {
+		if !attached {
+			attach { [weak self](error) in
+				if let error = error {
+					callback(error)
+				} else {
+					if let self = self {
+						self.joinRoom(withRoomId: roomId, username: username) { (error) in
+							callback(error)
+						}
+					} else {
+						callback(JanusResultError.codeErr(code: -1, desc: "Publish attach error"))
+					}
+				}
+			}
+			return
+		}
+		
+		setupPeerStream()
+		
+		super.joinRoom(withRoomId: roomId, username: username) { [weak self] (error) in
+			guard let self = self else { return }
+			if error == nil {
+				self.sendOffer()
+			}
+			callback(error)
+		}
+	}
+	
+	override func leaveRoom(callback: @escaping RoleLeaveRoomCallback) {
+		super.leaveRoom { [weak self] in
+			guard let self = self else { return }
```

Comment:
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/ClearKeep/ck-ios/blob/1ab8931f0f28255ebeb64e047f8b06b326be6a2d/Packages/ChatSecure/ChatSecure/Classes/WebRTCJanus/JanusRolePublish.swift#L165"></span>Override methods which only call super can be removed</td>
    </tr>
  </tbody>
</table>